new custom vpn section (#134)
NethServer/nethsecurity-docs: custom_openvpn_tunnel.rst

.. _custom_openvpn_tunnel-section:

=====================
Custom OpenVPN tunnel
=====================

This guide explains how to configure an OpenVPN client on NethSecurity using a configuration file (``myvpn.ovpn``) provided by a VPN service provider. 
The configuration ensures the VPN starts automatically when the firewall boots.

Prerequisites
-------------

- A valid OpenVPN configuration file (`myvpn.ovpn`) from your VPN provider.
- Access to the NethSecurity terminal via SSH.
- Basic familiarity with the UCI (Unified Configuration Interface) system in OpenWrt/NethSecurity.

> Additional Notes on CLI Configuration

Reviewer: ```suggestion
Additional notes on CLI configuration
```

-------------------------------------

- This procedure does not include any validation of the entered data. Therefore, it is intended to be performed by advanced users familiar with the NethSecurity environment and OpenVPN configurations.
- The VPN created using this method will not appear in the NethSecurity web interface and can only be managed through the command line interface (CLI).
- It is critical to avoid using the same name for a VPN created via the CLI and one configured through the NethSecurity web interface. Since there are no safeguards in place to prevent naming conflicts, such overlap may cause configuration issues.

For these reasons, caution and attention to detail are strongly advised when performing this procedure.

> Steps to Configure the VPN

Reviewer: ```suggestion
Steps to configure the VPN
```

--------------------------

> 1. Place the Configuration File in the Correct Directory

Reviewer: ```suggestion
1. Place the configuration file in the correct Directory
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1. Copy the ``myvpn.ovpn`` file to the directory ``/etc/openvpn/``. Use SCP or a similar tool to transfer the file: ::

    scp myvpn.ovpn root@<NethSecurity_IP>:/etc/openvpn/

> 2. Verify the permissions for the file, the output should look like this: ::
> 
>     ls -l /etc/openvpn/myvpn.ovpn
>     -rw-r--r--    1 root     root        <size> myvpn.ovpn
> 
> If it is not so then set the correct permissions: ::
> 
>     chmod 644 /etc/openvpn/myvpn.ovpn
>     chown root:root /etc/openvpn/myvpn.ovpn

Reviewer: ```suggestion
2. Make sure to set the correct file permissions: ::

    chmod 644 /etc/openvpn/myvpn.ovpn
    chown root:root /etc/openvpn/myvpn.ovpn
```

> 3. Create a New OpenVPN Client Configuration in UCI

Reviewer: ```suggestion
3. Create a new OpenVPN Client configuration in UCI
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1. Add a new OpenVPN section in the UCI database called *myvpn*, link the configuration file to this section and enable the VPN ::

    uci add openvpn openvpn
    uci rename openvpn.@openvpn[-1]='myvpn'
    uci set openvpn.myvpn.enabled='1'
    uci set openvpn.myvpn.config='/etc/openvpn/myvpn.ovpn'

2. Commit the changes to save the configuration: ::

    uci commit openvpn
   

> 4. Start the VPN Client Immediately

Reviewer: ```suggestion
4. Start the VPN client immediately
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
To start the VPN client without rebooting the system, run: ::

    /etc/init.d/openvpn restart

> This will restart all configured openvpn tunnels.

Reviewer: ```suggestion
This will restart all configured OpenVPN tunnels.
```

> 5. Verify the VPN is Running

Reviewer: ```suggestion
5. Verify the VPN is running
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Check the OpenVPN logs to confirm the connection: ::

    tail -f /var/log/messages | grep openvpn

You should see log entries indicating a successful connection.

.. note:: 

  - **File Name Consistency:** The configuration name ``myvpn`` must match the OpenVPN section name in UCI and the configuration file's location. If you change the name, ensure all references to ``myvpn`` in commands and filenames are updated.
  - **Automatic Startup:** By setting ``enabled='1'``, the VPN client will automatically start whenever the firewall boots.

Optional (1): Disable Automatic Startup
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If you want to prevent the VPN from starting automatically when the firewall boots, you can disable it using the following commands: ::

1. Disable the VPN in UCI: ::

    uci set openvpn.myvpn.enabled='0'
    uci commit openvpn

2. Restart the running VPN tunnels (cause it's disabled now, it will stop, while other tunnels will do a full restart): ::

>     /etc/init.d/openvpn stop

Reviewer: This is not correct: the proposed configuration completely disable the tunnel, not just at boot

> Optional (2): Configure Authentication Credentials

Reviewer: ```suggestion
Configure authentication credentials (optional)
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If the VPN requires a username and password, create an authentication file.

1. Create a file named ``/etc/openvpn/myvpn.auth`` (replace ``myvpn`` with the VPN name if different): ::

    vi /etc/openvpn/myvpn.auth

2. Add the following content, replacing ``frank`` and ``frank_password`` with your username and password: ::
                                    

>     user frank
>     password frank_password

Reviewer: ```suggestion
    frank
    frank_password
```

3. Save and set the correct permissions: ::

    chmod 600 /etc/openvpn/myvpn.auth
    chown root:root /etc/openvpn/myvpn.auth
                                    

> 4. Update the OpenVPN configuration file (``myvpn.ovpn``) to reference the authentication file. Open the file for editing: ::
>    
>     vi /etc/openvpn/myvpn.ovpn
> 
> 5. Add or modify the following line: ::
>                                     
>     auth-user-pass /etc/openvpn/myvpn.auth

Reviewer: ```suggestion
4. Update the OpenVPN configuration file (``myvpn.ovpn``) to reference the authentication file. ::
   
    echo "auth-user-pass /etc/openvpn/myvpn.auth" >>  /etc/openvpn/myvpn.ovpn

```

The output should resemble the following (example configuration name ``myvpn``): ::

    4913 ?        S      0:00 /usr/sbin/openvpn --syslog openvpn(myvpn) --status /var/run/openvpn.myvpn.status --cd /etc/openvpn --config myvpn.ovpn --up /usr/libexec/openvpn-hotplug up myvpn --down /usr/libexec/openvpn-hotplug down myvpn --route-up /usr/libexec/openvpn-hotplug route-up myvpn --route-pre-down /usr/libexec/openvpn-hotplug route-pre-down myvpn --script-security 2

Confirm the ``--config`` parameter points to the correct configuration file (e.g., ``myvpn.ovpn``).
Ensure all references (e.g., ``myvpn``) match your intended VPN configuration.

2. Check the OpenVPN logs to confirm the connection: ::

    tail -f /var/log/messages | grep openvpn

You should see log entries indicating a successful connection.

.. note:: 

  - **File Name Consistency:** The configuration name ``myvpn`` must match the OpenVPN section name in UCI and the configuration file's location. If you change the name, ensure all references to ``myvpn`` in commands and filenames are updated.
  - **Automatic Startup:** By setting ``enabled='1'``, the VPN client will automatically start whenever the firewall boots.

> Optional: Disable Automatic Startup

Reviewer: I'd move this section to the end.
```suggestion
Disable the tunnel
```

Author: > I'd move this section to the end.

Do you mean after the section: "Configure authentication credentials (optional)"
or aty the very end after the firewall section?

Reviewer: I'd say to the very end

Author: > I'd say to the very end

Section moved

                                    
    frank
    frank_password

3. Save and set the correct permissions: ::

    chmod 600 /etc/openvpn/myvpn.auth
    chown root:root /etc/openvpn/myvpn.auth
                                    
4. Update the OpenVPN configuration file (``myvpn.ovpn``) to reference the authentication file. ::
   
    echo "auth-user-pass /etc/openvpn/myvpn.auth" >>  /etc/openvpn/myvpn.ovpn

                                    
.. note:: 
                                    
  - **Authentication File:** When using an authentication file, ensure it has strict permissions (`600`) to protect sensitive information.

Configure the firewall to allow traffic for the VPN
-----------------------------------------------------

To enable traffic through the VPN, it is necessary to configure the firewall on NethSecurity. 
The best practice is to assign a fixed device name to the VPN, create a dedicated zone for the custom VPN, and associate the VPN device with that zone.

1. Fix the VPN device name
^^^^^^^^^^^^^^^^^^^^^^^^^^
To ensure the VPN device name remains consistent and avoids automatic assignment, it is crucial to fix the name in the OpenVPN configuration file. 

> Edit the file (``/etc/openvpn/myvpn.ovpn``) and include the following lines or modify them if they already exist (this example is made with a *routed* vpn): ::

Reviewer 2: ```suggestion
Edit the file (``/etc/openvpn/myvpn.ovpn``) to change "dev tun" to "dev tunmyvpn" and add the following line (this example is made with a *routed* vpn): ::
```

>     -dev tun

Reviewer 2: ```suggestion

```

>     +dev tunmyvpn

Reviewer 2: ```suggestion

```

>     +dev-type tun

Reviewer 2: ```suggestion
    dev-type tun
```

.. note:: 
                                    
  - **Authentication File:** When using an authentication file, ensure it has strict permissions (`600`) to protect sensitive information.

Configure the firewall to allow traffic for the VPN
-----------------------------------------------------

To enable traffic through the VPN, it is necessary to configure the firewall on NethSecurity. 
The best practice is to assign a fixed device name to the VPN, create a dedicated zone for the custom VPN, and associate the VPN device with that zone.

1. Fix the VPN device name
^^^^^^^^^^^^^^^^^^^^^^^^^^
To ensure the VPN device name remains consistent and avoids automatic assignment, it is crucial to fix the name in the OpenVPN configuration file. 
Edit the file (``/etc/openvpn/myvpn.ovpn``) and include the following lines or modify them if they already exist (this example is made with a *routed* vpn): ::

>     -dev tun

Author: yes

>     -dev tun

Author: Suggestions have been committed, a warning has beed added for interface's name maximum lenght.
